### PR TITLE
[hail] distinguish LZ4HC from LZ4Fast in buffer specs

### PR DESCRIPTION
**Pull request by danking**
Stacked on #7260 .

Also set wire and memory spec to LZ4Fast.

It was a bit annoying to make serialization work with a shared super class, so if anything looks funny down there that's probably why.

### CODE REVIEW

File: hail/src/main/scala/is/hail/io/BufferSpecs.scala [hunk 1]
```diff
@@ -99,23 +130,50 @@ trait BlockBufferSpec extends Spec {
...
 
-  def nativeOutputBufferType: String = s"LZ4OutputBlockBuffer<${ 4 + LZ4Utils.maxCompressedLength(blockSize) }, ${ child.nativeOutputBufferType }>"
+  def nativeOutputBufferType: String = s"$typeName<${ 4 + lz4.maxCompressedLength(blockSize) }, ${ child.nativeOutputBufferType }>"
 
   def nativeInputBufferType(inputStreamType: String): String =
-    s"LZ4InputBlockBuffer<${ 4 + LZ4Utils.maxCompressedLength(blockSize) }, ${ child.nativeInputBufferType(inputStreamType) }, $inputStreamType>"
+    s"$typeName<${ 4 + lz4.maxCompressedLength(blockSize) }, ${ child.nativeInputBufferType(inputStreamType) }, $inputStreamType>"
+}
+
+@deprecated("LZ4HCBlockBufferSpec is a drop-in replacement for this class", "Hail 0.2.24")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/BufferSpecs.scala]: just move this to `is.hail.compatibility`

File: hail/src/main/scala/is/hail/io/BufferSpecs.scala [hunk 2]
```diff
@@ -99,23 +130,50 @@ trait BlockBufferSpec extends Spec {
...
+  def buildOutputBuffer(out: OutputStream): OutputBlockBuffer = new LZ4OutputBlockBuffer(lz4, blockSize, child.buildOutputBuffer(out))
 
   def buildCodeInputBuffer(in: Code[InputStream]): Code[InputBlockBuffer] =
     Code.newInstance[LZ4InputBlockBuffer, Int, InputBlockBuffer](blockSize, child.buildCodeInputBuffer(in))
 
   def buildCodeOutputBuffer(out: Code[OutputStream]): Code[OutputBlockBuffer] =
     Code.newInstance[LZ4OutputBlockBuffer, Int, OutputBlockBuffer](blockSize, child.buildCodeOutputBuffer(out))
 
-  def nativeOutputBufferType: String = s"LZ4OutputBlockBuffer<${ 4 + LZ4Utils.maxCompressedLength(blockSize) }, ${ child.nativeOutputBufferType }>"
+  def nativeOutputBufferType: String = s"$typeName<${ 4 + lz4.maxCompressedLength(blockSize) }, ${ child.nativeOutputBufferType }>"
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/BufferSpecs.scala]: this method doesn't even seem to be used. Can we delete?

File: hail/src/main/scala/is/hail/io/BufferSpecs.scala [hunk 3]
```diff
@@ -1,38 +1,69 @@
...
-      LZ4BlockBufferSpec(32 * 1024,
+      LZ4HCBlockBufferSpec(32 * 1024,
         new StreamBlockBufferSpec)))
 
   val defaultUncompressed: BufferSpec = BlockingBufferSpec(32 * 1024,
     new StreamBlockBufferSpec)
 
   val unblockedUncompressed: BufferSpec = new StreamBufferSpec
 
+  val wireSpec: BufferSpec = LEB128BufferSpec(
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/io/BufferSpecs.scala]: I bet LEB is going to hurt us too. But let's do that experiment later.

File: hail/src/main/scala/is/hail/io/BufferSpecs.scala (hunk 2 above)
Comment (danking) [reply hail/src/main/scala/is/hail/io/BufferSpecs.scala]: I will give it a try

**Comment by danking** [comment]
ugh frig yeah it adds a new buffer spec.

**Comment by danking** [comment]
OK, I rebased, bumped the file format, and rebuilt each file. I bumped to 1.3.0 from 1.2.0 because it was backwards compatible new functionality.

**Comment by tpoterba** [comment]
I think we may need to rethink our compat tests at some point -- a few more bumps and that test will take minutes to run

**Pull request merged**